flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesJobGraphStoreWatcher.java

```diff
@@ -0,0 +1,148 @@
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	private JobGraphStore.JobGraphListener jobGraphListener;
+
+	@Nullable
+	private KubernetesWatch kubernetesWatch;
+
+	public KubernetesJobGraphStoreWatcher(FlinkKubeClient kubeClient, String configMapName) {
+		this.kubeClient = checkNotNull(kubeClient);
+		this.configMapName = checkNotNull(configMapName);
+	}
+
+	@Override
+	public void start(JobGraphStore.JobGraphListener jobGraphListener) {
+		this.jobGraphListener = checkNotNull(jobGraphListener);
+		kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+	}
+
+	@Override
+	public void stop() {
+		if (kubernetesWatch != null) {
+			kubernetesWatch.close();
+		}
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+		// This is used to get the difference between current and previous data. And then notify the listener.
+		final Set<JobID> previousJobIDs = new HashSet<>();
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			// The ConfigMap is created by KubernetesLeaderElectionDriver with empty data. We do not process this
+			// useless event.
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			handleConfigMapEvent(configMaps);
+		}
+
+		@Override
+		public void onDeleted(List<KubernetesConfigMap> configMaps) {
+			// The ConfigMap will be created again in the leader election service.
+		}
+
+		@Override
+		public void onError(List<KubernetesConfigMap> configMaps) {
+			LOG.error("Error while watching the configMap {}", configMapName);
+		}
+
+		@Override
+		public void handleFatalError(Throwable throwable) {
+			LOG.error("Error while watching the configMap {}", configMapName, throwable);
+		}
+
+		private void handleConfigMapEvent(List<KubernetesConfigMap> configMaps) {
+			final KubernetesConfigMap configMap = KubernetesUtils.checkConfigMaps(configMaps, configMapName);
+			final Set<JobID> currentJobIDs = getJobIDs(configMap);
+			notifyJobGraphListenChanges(currentJobIDs, previousJobIDs);
+			previousJobIDs.clear();
+			previousJobIDs.addAll(currentJobIDs);
+		}
+
+		private Set<JobID> getJobIDs(KubernetesConfigMap configMap) {
+			final Set<JobID> jobIDs;
+			if (configMap.getData() == null) {
+				jobIDs = Collections.emptySet();
+			} else {
+				jobIDs = configMap.getData().keySet().stream()
+					.filter(k -> k.startsWith(JOB_GRAPH_STORE_KEY_PREFIX))
+					.map(KubernetesUtils::jobGraphConfigMapKeyToJobID)
+					.collect(Collectors.toSet());
+			}
+			return jobIDs;
+		}
+
+		private void notifyJobGraphListenChanges(Set<JobID> currentJobIDs, Set<JobID> previousJobIDs) {
+			final Set<JobID> jobIDsToAdd = new HashSet<>();
+			final Set<JobID> jobIDsToRemove = new HashSet<>(previousJobIDs);
+
+			currentJobIDs.forEach(jobID -> {
```

Comment:
I would suggest to replace `currentJobIDs .forEach` with a foreach loop `for (JobID  jobID : currentJobIDs)`.